### PR TITLE
Plugin Bundle

### PR DESCRIPTION
**Pull request by mikeynap**
### PR checklist
- [x] DB CRUD
- [x] API CRUD
- [x] Models
- [x] Manager implementation
- [x] Functional Tests

### Change log
Added bundle for plugins, which groups multiple plugins together. Their inputs and outputs are sandboxed and aggregated.

This adds one layer of abstraction above services -- Bundles are to plugins as BundleService is to service. A bundle service provides the input needed to start the plugin bundle.

More information can be found here: https://rally1.rallydev.com/#/96594901848d/detail/userstory/113846303028

### CODE REVIEW

File: api/v1/convert/convert.go [hunk 1]
```diff
@@ -23,13 +23,65 @@ func FromModelPlugin(o model.Plugin) apiclient.Plugin {
 	}
 }
 
+func FromAPIBundle(o apiclient.Bundle) model.Bundle {
```
Comment (stephenrlouie) [inline api/v1/convert/convert.go]: convert test for bundles?

File: Makefile [hunk 2]
```diff
@@ -14,9 +14,14 @@ test:
 
 integration:
 	./tests/integration/setup.sh
-	go test -race -cover --ldflags '${EXTLDFLAGS}' $(PACKAGES_FOR_INTEGRATION)
+	go test -v  --ldflags '${EXTLDFLAGS}' $(PACKAGES_FOR_INTEGRATION)
```
Comment (stephenrlouie) [inline Makefile]: do our tests have a race condition?

File: Makefile [hunk 3]
```diff
@@ -14,9 +14,14 @@ test:
...
 	./tests/integration/shutdown.sh
 
+debug:
```
Comment (stephenrlouie) [inline Makefile]: whats this one for now?

File: api/v1/convert/convert.go [hunk 4]
```diff
@@ -23,13 +23,65 @@ func FromModelPlugin(o model.Plugin) apiclient.Plugin {
...
+		Input:  ToPluginIO(o.Input),
+		Output: ToPluginIO(o.Input),
+	}
+	for _, s := range o.Services {
+		b.Services = append(b.Services, FromModelService(s))
+	}
+	return b
+}
+
+func FromAPIBundleService(o apiclient.BundleService) model.BundleService {
```
Comment (stephenrlouie) [inline api/v1/convert/convert.go]: convert test for BundleService

File: api/v1/server/server.go [hunk 5]
```diff
@@ -121,6 +121,46 @@ func (s *Server) AddService(w http.ResponseWriter, r *http.Request) {
...
+	var apiObj apiclient.Bundle
+	if err := decoder.Decode(&apiObj); err != nil || apiObj.Name == "" {
+		http.Error(w, fmt.Sprintf("Invalid Bundle Provided: %v", err), http.StatusBadRequest)
+		return
+	}
+	modObj := convert.FromAPIBundle(apiObj)
+	var theID string
+	var err error
+	if theID, err = s.db.InsertBundle(modObj); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (stephenrlouie) [inline api/v1/server/server.go]: internal server error?

File: api/v1/server/server.go [hunk 6]
```diff
@@ -121,6 +121,46 @@ func (s *Server) AddService(w http.ResponseWriter, r *http.Request) {
...
+	var apiObj apiclient.BundleService
+	if err := decoder.Decode(&apiObj); err != nil || apiObj.Name == "" {
+		http.Error(w, fmt.Sprintf("Invalid BundleService Provided: %v", err), http.StatusBadRequest)
+		return
+	}
+	modObj := convert.FromAPIBundleService(apiObj)
+	var theID string
+	var err error
+	if theID, err = s.db.InsertBundleService(modObj); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (stephenrlouie) [inline api/v1/server/server.go]: internal server error?

File: api/v1/server/server.go [hunk 7]
```diff
@@ -153,6 +193,82 @@ func (s *Server) DeleteService(w http.ResponseWriter, r *http.Request) {
...
+func (s *Server) DeleteBundle(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	vars := mux.Vars(r)
+	id, ok := vars["bundle-id"]
+	if !ok || id == "" {
+		http.Error(w, "No Bundle ID Provided.", http.StatusBadRequest)
+		return
+	}
+	if err := s.db.DeleteBundle(id); err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
```
Comment (stephenrlouie) [inline api/v1/server/server.go]: internal server error? we could get an error for other reasons

File: api/v1/server/server.go [hunk 8]
```diff
@@ -153,6 +193,82 @@ func (s *Server) DeleteService(w http.ResponseWriter, r *http.Request) {
...
+func (s *Server) DeleteBundleService(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	vars := mux.Vars(r)
+	id, ok := vars["bundleservice-id"]
+	if !ok || id == "" {
+		http.Error(w, "No Bundle Service ID Provided.", http.StatusBadRequest)
+		return
+	}
+	if err := s.db.DeleteBundleService(id); err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
```
Comment (stephenrlouie) [inline api/v1/server/server.go]: internal server error?

File: Makefile (hunk 2 above)
Comment (mikeynap) [inline Makefile]: I don't give a fuck if my function test has a race condition. It makes line debugging impossible, and it slows it down.

Comment (mikeynap) [inline Makefile]: Race conditions should be caught by the unit tests.

File: api/v1/server/server.go [hunk 9]
```diff
@@ -359,8 +526,84 @@ func (s *Server) UpdateService(w http.ResponseWriter, r *http.Request) {
...
+		var err error
+		if b2Update, err = s.db.GetBundle(id); err != nil {
+			return http.StatusNotFound, err
+		}
+		updatedB := convert.FromAPIBundle(apiObj)
+		updatedB.MetaImpl = b2Update.MetaImpl
+		updatedB.Set("Dirty", "true")
+		updatedB.Id = b2Update.Id
+		if err = s.db.UpdateBundle(updatedB); err != nil {
+			return http.StatusBadRequest, err
```
Comment (stephenrlouie) [inline api/v1/server/server.go]: internal server error?

File: api/v1/server/server.go [hunk 10]
```diff
@@ -359,8 +526,84 @@ func (s *Server) UpdateService(w http.ResponseWriter, r *http.Request) {
...
+		var err error
+		if b2Update, err = s.db.GetBundleService(id); err != nil {
+			return http.StatusNotFound, err
+		}
+		updatedB := convert.FromAPIBundleService(apiObj)
+		updatedB.MetaImpl = b2Update.MetaImpl
+		updatedB.Set("Dirty", "true")
+		updatedB.Id = b2Update.Id
+		if err = s.db.UpdateBundleService(updatedB); err != nil {
+			return http.StatusBadRequest, err
```
Comment (stephenrlouie) [inline api/v1/server/server.go]: internal server error?

File: api/v1/convert/convert.go (hunk 1 above)
Comment (mikeynap) [inline api/v1/convert/convert.go]: I mean I could... it just wraps the already tested functions. Seems like a lot of typing.

File: Makefile (hunk 3 above)
Comment (mikeynap) [inline Makefile]: It builds with extra debugging information enabled for GDB.

File: api/v1/convert/convert.go (hunk 4 above)
Comment (mikeynap) [inline api/v1/convert/convert.go]: See above comment.

File: api/v1/server/server.go (hunk 7 above)
Comment (mikeynap) [inline api/v1/server/server.go]: Good question. Delete returns not found if the ID is not found.

File: manager/manager.go [hunk 11]
```diff
@@ -114,82 +113,124 @@ func (m *manager) Start() error {
...
 
 // Stop implements the service to stop the manager
 func (m *manager) Stop() {
 	close(m.quit)
 }
 
+// Fills in bundle FROMs, adds all services.
+// TODO: Add After/When/Before etc.
+func (m *manager) processBundles() {
+	bundleservices, err := m.db.QueryBundleServices(model.BundleService{})
```
Comment (stephenrlouie) [inline manager/manager.go]: could we have a more selective query instead of ALL bundles?

File: manager/manager.go [hunk 12]
```diff
@@ -461,8 +497,9 @@ func (m *manager) collectInput(svc *model.Service, plugin model.Plugin) (bool, e
 				}
 
 				qService := model.Service{
-					PluginName: pts[0],
-					Name:       pts[1],
+					//PluginName:    pts[0],
```
Comment (stephenrlouie) [inline manager/manager.go]: ?

Comment (mikeynap) [inline manager/manager.go]: Yeah... we all gotta chat about dis one.

File: manager/manager.go (hunk 11 above)
Comment (mikeynap) [inline manager/manager.go]: Yeah, I'll query on the "unprocessed" ones. Good catch.

File: manager/manager.go [hunk 13]
```diff
@@ -611,14 +647,70 @@ func (m *manager) pluginFor(svc model.Service) (model.Plugin, error) {
...
+			}
+			if !set {
+				return ErrInputsNotSatisfied
+			}
+			hasBundle = true
+		} else {
+			hasNonBundle = true
+		}
+	}
+	if value != nil && hasBundle && hasNonBundle {
```
Comment (stephenrlouie) [inline manager/manager.go]: so bundle input cannot be taken from a bundle and non-bundle at the same time? Why?

File: manager/manager_test.go [hunk 14]
```diff
@@ -947,22 +947,21 @@ func expectCollectInput(t *testing.T, m *manager, db *mock_database.MockDatabase
 				if _, ok := sources[strings.Join(pts[:2], ".")]; !ok {
 					// Expect empty GET
 					db.EXPECT().QueryServices(model.Service{
-						PluginName: pts[0],
-						Name:       pts[1],
+						//PluginName: pts[0],
```
Comment (stephenrlouie) [inline manager/manager_test.go]: ?

File: manager/manager_test.go [hunk 15]
```diff
@@ -947,22 +947,21 @@ func expectCollectInput(t *testing.T, m *manager, db *mock_database.MockDatabase
...
+						Name: pts[1],
 					}).Return([]model.Service{}, nil)
 					continue
 				}
 				// Expect GET
 				db.EXPECT().QueryServices(model.Service{
-					PluginName: pts[0],
-					Name:       pts[1],
+					//PluginName: pts[0],
```
Comment (stephenrlouie) [inline manager/manager_test.go]: ?

File: tests/integration/manager/test/test_insert_bundle.sh [hunk 16]
```diff
@@ -0,0 +1,14 @@
+#! /bin/bash
+
+curl localhost:2481/v1/plugins -d @test_plugin_foo.json
+curl localhost:2481/v1/plugins -d @test_plugin_bar.json
+curl localhost:2481/v1/services -d @test_service_foo.json
+curl localhost:2481/v1/services -d @test_service_foo2.json
+curl localhost:2481/v1/services -d @test_service_bar.json
+
+
+#curl localhost:2481/v1/plugins -d @test_plugin_converter.json
```
Comment (stephenrlouie) [inline tests/integration/manager/test/test_insert_bundle.sh]: ?

File: tests/integration/manager/test/test_plugin_drmer.json [hunk 17]
```diff
@@ -0,0 +1,22 @@
+{
+	"id": "test-plugin-drmer",
```
Comment (stephenrlouie) [inline tests/integration/manager/test/test_plugin_drmer.json]: can we rename to generic foo/bar names. not use case specific

File: tests/integration/manager/test/test_plugin_converter.json [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    }
+	],
+	"output": [
+		{
+			"name": "url",
+			"type": "url"
+		}
+	],
+	"config": {
+		"image": "mikeynap/ditto:latest"
```
Comment (stephenrlouie) [inline tests/integration/manager/test/test_plugin_converter.json]: lets push the examples to the sigmadev dockerhub

File: api/v1/client/bundle.go [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+ * This outlines version 1 of the Mu Sigma API. All requests should be prefixed with the `v1` basePath. 
+ *
+ * OpenAPI spec version: 1.0
+ * 
+ * Generated by: https://github.com/swagger-api/swagger-codegen.git
+ */
+
+package apiclient
+
+type Bundle struct {
```
Comment (meganokeefe) [inline api/v1/client/bundle.go]: Thoughts on renaming `Bundle` to `BundlePlugin`? To make it clear that a bundle is essentially a plugin. Similar to the naming of `BundleService`

File: api/v1/convert/convert.go (hunk 1 above)
Comment (meganokeefe) [inline api/v1/convert/convert.go]: do it. improve that test coverage.

File: api/v1/client/service_io.go [hunk 20]
```diff
@@ -1,22 +1,23 @@
...
-
 	Name string `json:"name,omitempty"`
 
 	Type_ string `json:"type,omitempty"`
 
 	Value interface{} `json:"value,omitempty"`
 
 	From interface{} `json:"from,omitempty"`
+
+	Actual interface{} `json:"actual,omitempty"`
```
Comment (meganokeefe) [inline api/v1/client/service_io.go]: what is `Actual` used for?

File: api/v1/convert/convert.go [hunk 21]
```diff
@@ -128,10 +179,11 @@ func ToServiceDataItems(o []apiclient.ServiceIo) []model.ServiceDataItem {
...
 		r = append(r, model.ServiceDataItem{
-			Name:  io.Name,
-			Type:  io.Type_,
-			From:  interfaceToStringArray(io.From),
-			Value: interfaceToStringArray(io.Value),
+			Name:   io.Name,
+			Type:   io.Type_,
+			From:   interfaceToStringArray(io.From),
+			Value:  interfaceToStringArray(io.Value),
+			Actual: interfaceToStringArray(io.Value), // Do I do this??????????
```
Comment (meganokeefe) [inline api/v1/convert/convert.go]: remove comment

File: api/v1/server/server.go [hunk 22]
```diff
@@ -114,13 +137,54 @@ func (s *Server) AddService(w http.ResponseWriter, r *http.Request) {
...
+	defer r.Body.Close()
+	var apiObj apiclient.BundleService
+	if err := decoder.Decode(&apiObj); err != nil || apiObj.Name == "" {
+		http.Error(w, fmt.Sprintf("Invalid BundleService Provided: %v", err), http.StatusBadRequest)
+		return
+	}
+	modObj := convert.FromAPIBundleService(apiObj)
+	var theID string
+	var err error
+	modObj.Set("processed", "false")
```
Comment (meganokeefe) [inline api/v1/server/server.go]: what is this line for

File: tests/integration/manager/test/test_plugin_drmer.json (hunk 17 above)
Comment (mikeynap) [inline tests/integration/manager/test/test_plugin_drmer.json]: No.

File: tests/integration/manager/test/test_plugin_converter.json (hunk 18 above)
Comment (mikeynap) [inline tests/integration/manager/test/test_plugin_converter.json]: already did -- will update the code.

File: tests/integration/manager/manager_test.go [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+	cmd := exec.Command("sh", script+"/test_insert_bundle.sh")
+	flog.Println("Inserting bundle...")
+	cmd.Dir = script
+	if err := cmd.Run(); err != nil {
+		flog.Println(err)
+		os.Exit(1)
+	}
+	cli = apiclient.NewDefaultApiWithBasePath("http://localhost:2481/v1")
+	flog.Println("Testing for populated input/output values...")
+	/*if err := testBundleServices(t); err != nil {
```
Comment (meganokeefe) [inline tests/integration/manager/manager_test.go]: uncomment this?

File: api/v1/client/service_io.go (hunk 20 above)
Comment (mikeynap) [inline api/v1/client/service_io.go]: Adam uses this to store the value of the service -- i.e. the outputs.

File: api/v1/server/server.go (hunk 22 above)
Comment (mikeynap) [inline api/v1/server/server.go]: to mark a bundle service as "processed" -- meaning the services are running

File: tests/integration/manager/manager_test.go (hunk 23 above)
Comment (mikeynap) [inline tests/integration/manager/manager_test.go]: yeah idk how that got pushed lol my b.

File: tests/integration/manager/manager_test.go [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+		t.Error("Failed testing bundleservices:", err)
+		t.FailNow()
+	}*/
+	if err := testServices(t); err != nil {
```
Comment (meganokeefe) [inline tests/integration/manager/manager_test.go]: This isn't related to bundles, it's more related to manager/serviceOps but: 

I'm not sure `/var/lib/sigma` is the best directory for inputs and outputs.
On my machine, `/var/lib` is default owned by root. And you can't `sudo make integration`. It has to just work. So i'd have to `chown` etc. on `/var/lib.`

And so initially, running `make integration` for the bundle tests, I got this, with no context: 

```
➜  sigma git:(bundle) make integration
./tests/integration/setup.sh
Integration tests dependency set up
go test -v  --ldflags '' github.com/sigma-dev/sigma/tests/integration/manager github.com/sigma-dev/sigma/tests/integration/task/docker
=== RUN   TestManager
time="2017-05-09T11:48:13-04:00" level=info msg="Starting Sigma..."
time="2017-05-09T11:48:14-04:00" level=info msg="Inserting bundle..."
time="2017-05-09T11:48:14-04:00" level=info msg="Testing for populated input/output values..."
time="2017-05-09T11:48:15-04:00" level=error msg="Could not start operation 08bf6732-187a-41c7-a30f-0dee2abf9790-operation" error="mkdir /var/lib/sigma/08bf6732-187a-41c7-a30f-0dee2abf9790: permission denied" file=manager.go line=291
time="2017-05-09T11:48:15-04:00" level=warning msg="Could not remove container 08bf6732-187a-41c7-a30f-0dee2abf9790" error="Error response from daemon: No such container: 08bf6732-187a-41c7-a30f-0dee2abf9790" file=manager.go line=293
time="2017-05-09T11:48:15-04:00" level=error msg="Could not start operation 7fe42d23-bf25-42c6-837a-b87e3e2feac9-operation" error="mkdir /var/lib/sigma/7fe42d23-bf25-42c6-837a-b87e3e2feac9: permission denied" file=manager.go line=291
time="2017-05-09T11:48:15-04:00" level=warning msg="Could not remove container 7fe42d23-bf25-42c6-837a-b87e3e2feac9" error="Error response from daemon: No such container: 7fe42d23-bf25-42c6-837a-b87e3e2feac9" file=manager.go line=293
time="2017-05-09T11:48:17-04:00" level=error msg="Could not start operation 08bf6732-187a-41c7-a30f-0dee2abf9790-operation" error="mkdir /var/lib/sigma/08bf6732-187a-41c7-a30f-0dee2abf9790: permission denied" file=manager.go line=291
time="2017-05-09T11:48:17-04:00" level=warning msg="Could not remove container 08bf6732-187a-41c7-a30f-0dee2abf9790" error="Error response from daemon: No such container: 08bf6732-187a-41c7-a30f-0dee2abf9790" file=manager.go line=293
time="2017-05-09T11:48:17-04:00" level=error msg="Could not start operation 7fe42d23-bf25-42c6-837a-b87e3e2feac9-operation" error="mkdir /var/lib/sigma/7fe42d23-bf25-42c6-837a-b87e3e2feac9: permission denied" file=manager.go line=291
```

File: tests/integration/manager/manager_test.go [hunk 25]
```diff
@@ -0,0 +1,229 @@
...
+		t.Log("Failed testing services:", err)
+		t.FailNow()
+	}
+	t.Log("Success!")
+	return
+}
+
+func ActuallyEqual(i interface{}, val []string) bool {
```
Comment (meganokeefe) [inline tests/integration/manager/manager_test.go]: Nit, rename this to something more descriptive, eg. `InterfaceEqualsSlice`

File: tests/integration/manager/manager_test.go [hunk 26]
```diff
@@ -0,0 +1,229 @@
...
+		ival, ok := ia[i].(string)
+		if !ok || ival != v {
+			return false
+		}
+	}
+	return true
+}
+
+func testBundleServices(t *testing.T) error {
+	bundle, _, err := cli.Getbundle("WorkflowUserstoryBundle")
```
Comment (meganokeefe) [inline tests/integration/manager/manager_test.go]: this should be `GetBundle`, not `Getbundle`, yes?

File: tests/integration/manager/manager_test.go [hunk 27]
```diff
@@ -0,0 +1,229 @@
...
+	}
+
+	bs, _, err := cli.GetBundleService("WorkflowUserstoryBundleService")
+	if err != nil {
+		return err
+	}
+	if bs == nil {
+		return errors.New("No bundle service...")
+	}
+
```
Comment (meganokeefe) [inline tests/integration/manager/manager_test.go]: nit remove newline

File: tests/integration/manager/manager_test.go [hunk 28]
```diff
@@ -0,0 +1,229 @@
...
+type test struct {
+	serviceName string
+	expInput    map[string][]string
+	expOutput   map[string][]string
+	passed      bool
+}
+
+func runTests(tt *testing.T, tests []*test) error {
+	success := false
+	for i := 0; i < 5; i++ {
```
Comment (meganokeefe) [inline tests/integration/manager/manager_test.go]: why hardcoding iterations to 5? reasoning?

File: tests/integration/manager/manager_test.go [hunk 29]
```diff
@@ -0,0 +1,229 @@
...
+				continue
+			}
+			passed = false
+			break
+		}
+		if passed {
+			success = true
+			break
+		}
+		time.Sleep(5 * time.Second)
```
Comment (meganokeefe) [inline tests/integration/manager/manager_test.go]: is there any way, in the future, that we could get around hardcoding a `sleep` here? or do some kind of try-retry status check? 

this goes along with, why are we running these tests 5 times over? is it for load purposes?

"je suis confus"

File: tests/integration/manager/manager_test.go [hunk 30]
```diff
@@ -0,0 +1,229 @@
...
+	}
+	if !success {
+		return errors.New("Services not correctly populated in time.")
+	}
+	return nil
+}
+
+func actual(io []apiclient.ServiceIo, k string) interface{} {
```
Comment (meganokeefe) [inline tests/integration/manager/manager_test.go]: comment this function or rename it to something descriptive. 

it seems like a find-in-slice function.

File: database/nulldb.go [hunk 31]
```diff
@@ -247,6 +376,42 @@ func (n *NullDB) QueryServicesIds(o model.Service) ([]string, error) {
...
+	n.Lock()
+	defer n.Unlock()
+	r := make([]model.Bundle, 0)
+	mdMatch := true
+	for _, v := range n.Bundles {
+		for k, mdv := range o.Metadata {
+			mdMatch = mdMatch && mdv == v.Metadata[k]
+		}
+
+		if mdMatch && (o.Id == "" || o.Id == v.Id) && (o.Name == "" || o.Name == v.Name) {
```
Comment (meganokeefe) [inline database/nulldb.go]: NIT, should have full query params here, not just for Id and Name

Comment (meganokeefe) [inline database/nulldb.go]: maybe not worth it for this PR, but might be valuable at some point to add a `stringInSlice` type helper that would allow you to query NullDB on `Inputs`, `Outputs`, and `Services` for Bundles (and BundleServices).

File: database/couchbase.go [hunk 32]
```diff
@@ -430,7 +523,7 @@ func isClean(s string) bool {
 	return reg.MatchString(s)
 }
 
-func (cb Couchbase) makeQuery(intf interface{}, idOnly bool) (string, map[string]interface{}, error) {
+func (cb Couchbase) makeQuery(obj interface{}, idOnly bool) (string, map[string]interface{}, error) {
```
Comment (meganokeefe) [inline database/couchbase.go]: why did you rename `intf` to `obj`

File: api/v1/server/server.go [hunk 33]
```diff
@@ -147,12 +211,88 @@ func (s *Server) DeleteService(w http.ResponseWriter, r *http.Request) {
...
+
+func (s *Server) GetBundleService(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	vars := mux.Vars(r)
+	id, ok := vars["bundleservice-id"]
+	if !ok || id == "" {
+		http.Error(w, "No Bundle ID Provided.", http.StatusBadRequest)
+		return
+	}
+	var modObj model.BundleService
```
Comment (meganokeefe) [inline api/v1/server/server.go]: are these two variable declarations needed?

File: api/v1/server/server.go [hunk 34]
```diff
@@ -276,6 +416,57 @@ func (s *Server) GetServices(w http.ResponseWriter, r *http.Request) {
...
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	queries := r.URL.Query()
+	bundlename := queries.Get("bundle")
+
+	apiObjs := make([]apiclient.BundleService, 0)
+	var modObjs []model.BundleService
+	var err error
+
+	obj := model.BundleService{
+		Bundle: bundlename,
```
Comment (meganokeefe) [inline api/v1/server/server.go]: shouldn't `obj` be an empty `BundleService`? if you want to get All? why are you adding `bundlename` here?

File: database/couchbase.go [hunk 35]
```diff
@@ -27,6 +27,8 @@ const (
 	pluginType           = "PLUGIN"
 	serviceType          = "SERVICE"
 	serviceOperationType = "SERVICE_OPERATION"
+	bundleType           = "BUNDLE_TYPE"
```
Comment (meganokeefe) [inline database/couchbase.go]: nit, for consistency these should be `BUNDLE` and `BUNDLE_SERVICE`, no `_TYPE`.

File: database/couchbase.go [hunk 36]
```diff
@@ -439,88 +532,84 @@ func (cb Couchbase) makeQuery(intf interface{}, idOnly bool) (string, map[string
...
 
-		for k, v := range op.Metadata {
-			if isClean(k) {
-				query, isFirst = queryHelper(isFirst, query, fmt.Sprintf("Metadata.%s=$%s", k, strings.ToLower(k)))
-				params[strings.ToLower(k)] = v
-			} else {
-				log.Warn("Non-sanitary key value for query %s", k)
-			}
+	case model.Bundle:
+		if obj.Name != "" {
```
Comment (meganokeefe) [inline database/couchbase.go]: both Bundle and BundleService queries should support querying by `Id`.

File: database/couchbase.go [hunk 37]
```diff
@@ -559,6 +651,8 @@ func replaceErr(err error) error {
 		return ErrCas
 	case gocb.ErrTimeout:
 		return ErrTimeout
+	case gocb.ErrIndexNotFound, gocb.ErrKeyNotFound:
```
Comment (meganokeefe) [inline database/couchbase.go]: why are we returning less specific errors and collapsing these? I feel like we should report the `gocb` errors as is, for ease of debug?

File: api/v1/convert/convert.go (hunk 4 above)
Comment (mikeynap) [inline api/v1/convert/convert.go]: go FMT yourself.

File: manager/manager.go [hunk 38]
```diff
@@ -611,14 +646,70 @@ func (m *manager) pluginFor(svc model.Service) (model.Plugin, error) {
...
+	plugins, err := m.db.QueryPlugins(qPlugin)
+	if err != nil {
 		return qPlugin, err
-	} else {
-		if len(plugins) != 1 {
-			return qPlugin, ErrPluginNotFound
-		}
-		plugin = plugins[0]
 	}
+	if len(plugins) != 1 {
```
Comment (meganokeefe) [inline manager/manager.go]: there could be an off chance where the query -> `len > 1` in which case that's a different error. maybe change to `len(plugins) == 0`

File: manager/manager.go [hunk 39]
```diff
@@ -611,14 +646,70 @@ func (m *manager) pluginFor(svc model.Service) (model.Plugin, error) {
...
+
+	// Get plugin
+	qBundle := model.Bundle{
+		Name: bs.Bundle,
+	}
+	bundles, err := m.db.QueryBundles(qBundle)
+	if err != nil {
+		return qBundle, err
+	}
+	if len(bundles) != 1 {
```
Comment (meganokeefe) [inline manager/manager.go]: see above comment, check if `len() == 0` instead? queries can theoretically turn up more than 1 bundle. etc.

File: manager/manager.go [hunk 40]
```diff
@@ -611,14 +646,70 @@ func (m *manager) pluginFor(svc model.Service) (model.Plugin, error) {
...
+		return qBundle, ErrBundleNotFound
+	}
+
+	return bundles[0], nil
+}
+
+// Assumes key in bundle's namespace is bundle.key
+//  len(in.From) > 1 --> isPrefix("bundle.", in.From[i]) ∀i ∈ [0,len(in.From))
```
Comment (meganokeefe) [inline manager/manager.go]: jeez, mike.

**Comment by meganokeefe** [comment]
ya got merge conflicts.

File: tests/integration/manager/manager_test.go (hunk 26 above)
Comment (mikeynap) [inline tests/integration/manager/manager_test.go]: shit yes. AHHHH DAMN YOU SWAGGER.

File: tests/integration/manager/manager_test.go (hunk 28 above)
Comment (mikeynap) [inline tests/integration/manager/manager_test.go]: none really

Comment (mikeynap) [inline tests/integration/manager/manager_test.go]: seems to be "long enough"

Comment (mikeynap) [inline tests/integration/manager/manager_test.go]: gives 30 seconds

File: tests/integration/manager/manager_test.go (hunk 29 above)
Comment (mikeynap) [inline tests/integration/manager/manager_test.go]: It's because the manager polls and takes it's time. 
It has to add the service, run the services, collect the output, wait for the service's dependencies to finish etc.

So this is waiting for those to be updated in the db... there's no other way to do this. This is saying "hm, there's no output yet, lets wait for it to finish".

File: api/v1/server/server.go (hunk 34 above)
Comment (mikeynap) [inline api/v1/server/server.go]: bundle name is an optional url-query parameter.

File: manager/manager.go (hunk 40 above)
Comment (mikeynap) [inline manager/manager.go]: If they can't read the math, get outta my kitchen.
but yeah sure.

File: manager/manager.go [hunk 41]
```diff
@@ -681,6 +681,9 @@ func (m *manager) bundleFor(bs model.BundleService) (model.Bundle, error) {
 
 // Assumes key in bundle's namespace is bundle.key
 //  len(in.From) > 1 --> isPrefix("bundle.", in.From[i]) ∀i ∈ [0,len(in.From))
+// i.e. If there are multiple froms, either all or none should be 'bundle.' Froms.
+// a mixture of From: [bundle.var, Plugin.Service.var] is not aloud.
```
Comment (meganokeefe) [inline manager/manager.go]: aloud -> allowed

File: manager/manager.go [hunk 42]
```diff
@@ -681,6 +681,9 @@ func (m *manager) bundleFor(bs model.BundleService) (model.Bundle, error) {
...
+// If this is needed, create a seperate input item for each.
```
Comment (meganokeefe) [inline manager/manager.go]: separate

File: tests/integration/manager/manager_test.go (hunk 25 above)
Comment (adamschaub) [inline tests/integration/manager/manager_test.go]: +1

File: database/couchbase.go (hunk 37 above)
Comment (adamschaub) [inline database/couchbase.go]: It's for interoperability between database implementations. Manager just checks for a `database.ErrNotFound`, regardless of what database it is using.

Comment (adamschaub) [inline database/couchbase.go]: Although we could provide our own database Error interface that also stores the original error.

File: api/v1/server/server.go (hunk 5 above)
Comment (adamschaub) [inline api/v1/server/server.go]: Could be either.

File: api/v1/server/server.go [hunk 43]
```diff
@@ -276,6 +392,57 @@ func (s *Server) GetServices(w http.ResponseWriter, r *http.Request) {
...
+func (s *Server) GetBundles(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	apiObjs := make([]apiclient.Bundle, 0)
+	var modObjs []model.Bundle
+	var err error
+
+	obj := model.Bundle{}
+
+	if modObjs, err = s.db.QueryBundles(obj); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (adamschaub) [inline api/v1/server/server.go]: internal server error?
